### PR TITLE
Do not complain about non-existing fields after parse recovery

### LINKED ISSUE #57361: Elide use of "non-existing" fields for enum variant with parse errors
On parse error of the fields of an enum variant, we also emit an error about "non-existing" fields when using the variant. We have no way of actually knowing wether those fields exist, we should elide the second error:

```
error: expected `:`, found `,`
   --> src/librustc/traits/mod.rs:228:37
    |
228 |     MatchExpressionArmPattern { span, ty: Ty<'tcx> },
    |                                     ^ expected `:`

error[E0026]: variant `traits::ObligationCauseCode::MatchExpressionArmPattern` does not have fields named `span`, `ty`
   --> src/librustc/infer/error_reporting/mod.rs:464:62
    |
464 |             ObligationCauseCode::MatchExpressionArmPattern { span, ty } => {
    |                                                              ^^^^  ^^ variant `traits::ObligationCauseCode::MatchExpressionArmPattern` does not have these fields
```

### PR DESCRIPTION
**Pull request by estebank**
When failing to parse struct-like enum variants, the ADT gets recorded
as having no fields. Record that we have actually recovered during
parsing of this variant to avoid complaing about non-existing fields
when actually using it.

Fix #57361.

### CODE REVIEW

File: src/test/ui/parser/recovered-struct-variant.stderr [hunk 1]
```diff
@@ -0,0 +1,8 @@
+error: expected `:`, found `,`
+  --> $DIR/recovered-struct-variant.rs:2:10
+   |
+LL |     A { a, b: usize }
+   |          ^ expected `:`
+
+error: aborting due to previous error
```
Comment (estebank) [inline src/test/ui/parser/recovered-struct-variant.stderr]: For comparison, the current output is

```
error: expected `:`, found `,`
 --> src/main.rs:2:10
  |
2 |     A { a, b: usize }
  |          ^ expected `:`

error[E0559]: variant `Foo::A` has no field named `a`
 --> src/main.rs:8:22
  |
8 |     let f = Foo::A { a:3, b: 4};
  |                      ^ `Foo::A` does not have this field

error[E0559]: variant `Foo::A` has no field named `b`
 --> src/main.rs:8:27
  |
8 |     let f = Foo::A { a:3, b: 4};
  |                           ^ `Foo::A` does not have this field

error[E0026]: variant `Foo::A` does not have fields named `a`, `b`
  --> src/main.rs:11:17
   |
11 |         Foo::A {a, b} => {}
   |                 ^  ^ variant `Foo::A` does not have these fields

error: aborting due to 4 previous errors
```

File: src/librustc/hir/mod.rs [hunk 2]
```diff
@@ -2173,7 +2173,7 @@ impl StructField {
 /// Id of the whole struct lives in `Item`.
 #[derive(Clone, RustcEncodable, RustcDecodable, Debug, HashStable)]
 pub enum VariantData {
-    Struct(HirVec<StructField>, HirId),
+    Struct(HirVec<StructField>, HirId, bool),
```
Comment (petrochenkov) [inline src/librustc/hir/mod.rs]:
```suggestion
    Struct(HirVec<StructField>, HirId, /*recovered*/ bool),
```

File: src/libsyntax/ast.rs [hunk 3]
```diff
@@ -2133,7 +2133,7 @@ pub enum VariantData {
     /// Struct variant.
     ///
     /// E.g., `Bar { .. }` as in `enum Foo { Bar { .. } }`.
-    Struct(Vec<StructField>, NodeId),
+    Struct(Vec<StructField>, NodeId, bool),
```
Comment (petrochenkov) [inline src/libsyntax/ast.rs]:
```suggestion
    Struct(Vec<StructField>, NodeId, /*recovered*/ bool),
```

File: src/libsyntax/parse/parser.rs [hunk 4]
```diff
@@ -6898,12 +6902,14 @@ impl<'a> Parser<'a> {
         }
     }
 
-    fn parse_record_struct_body(&mut self) -> PResult<'a, Vec<StructField>> {
+    fn parse_record_struct_body(&mut self) -> PResult<'a, (Vec<StructField>, bool)> {
```
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]:
```suggestion
    fn parse_record_struct_body(&mut self) -> PResult<'a, (Vec<StructField>, /*recovered*/ bool)> {
```
:slightly_smiling_face:

**Comment by petrochenkov** [comment]
r=me with the comments added

**Pull request merged**